### ISSUE TITLE
High memory & CPU usage after running for awhile

### ISSUE THREAD

**Issue by beyang** [sourcegraph/go-langserver#178]
Enabled the langserver and did a few hovers for testing. Then vscode was in the background. After a few minutes the langserver was using 5.6GB of memory (machine only has 4 GB RAM).

**Comment by keegancsmith** [comment]
Related to https://github.com/Microsoft/vscode-go/issues/853

So the issue is our language server resource usage is currently tuned for a server environment were we have a premise of lots of RAM. This has been presenting us issues in prod as well, so rewriting our hover and j2d to be smarter with resources needs to be done. It is almost certainly related to using the simple go loader interface and/or naive caching of its output. Add in our conservative cache invalidation policy, and we create a lot of garbage and re-work things out often in an editor environment.

I'll see if there is any short term wins here, but we probably need to make some larger changes to get good at this.

**Comment by vanackere** [comment]
Same issue for me: even with 16GB of memory I had to stop using go-langserver in vscode because it was frequently using all memory, making the computer unusable and triggering the OOM-Killer...

**Comment by vanackere** [comment]
Since I can reliably cause the langserver to use all memory on my private code base., how can I help debug this issue (or at least find where most of the ressource are spent) ? Any profiling tool to run ? If so, how ?

**Comment by ramya-rao-a** [comment]
@keegancsmith Any updates on this?

I have pushed a change for adding the ability to pass `-trace` and `-logfile` flags to the language server (See https://github.com/Microsoft/vscode-go/commit/50e5c8acd0553e04572b221924149b1b2f7926df)

Is there anything else you'd like to see users able to do?

**Comment by OneOfOne** [comment]
I hate to do this but any news on this? this is kind of a major issue.

I have 32gb of ram and the oom killer triggered twice in the last 4 hours, gocode is broken on go git.

**Comment by beyang** [comment]
@OneOfOne Apologies for the delay. We're still looking into it, but have had to prioritize a few other features for Go in the past 2 weeks. I'll update this issue when we have a patch targeting this issue. Appreciate the feedback and thanks for trying it out!

**Comment by mgleason** [comment]
+1 Same issue here (on MacOS).  Seems to grow to consume all available memory and then keeps going consuming swap as the system slows to a crawl.   

FYI - restarted "go-langserver", it re-spawned consuming 183.6MB of space.  Added a single space to a .go file and saved - memory consumed jumped to 576.0MB.  Changed a letter from lower to uppercase, saved, changed it back to lowercase, saved and memory was at 1.09GB.  Will try restarting next...

**Comment by beyang** [comment]
Thanks for the feedback @mgleason. I'll be targeting the local machine use case in the next couple of weeks, so I hope to have a patch out soon. In the meantime, thanks for trying it out and posting the feedback!

**Comment by keegancsmith** [comment]
A detailed comment for someone who tackles this problem.

TLDR: Rewrite everything in loader.go so we work more like the go compiler. https://sourcegraph.com/github.com/sourcegraph/go-langserver@ee6a3562/-/blob/langserver/loader.go

Currently this language server is optimized for the server environment on sourcegraph.com. The tradeoffs which most negatively effect desktop users are:
* Many shared read-only workspaces in a single process.
* Server environment has lots of RAM.

When we receive a request for `my/pkg/a.go`, we use `golang.org/x/tools/go/loader` to typecheck `my/pkg`. We only typecheck `my/pkg`, but the bulk of the time is spent on the transitive dependencies. https://sourcegraph.com/github.com/sourcegraph/go-langserver@ee6a3562/-/blob/langserver/loader.go#L233

Our caching for this is only to cache the output of loader for `my/pkg`. All the intermediate work it does (on the transitive dependencies) is not cached at all. When we receive an edit we just throw that cache out. This works alright for sourcegraph.com since when a user is browsing a file we never invalidate the cache and we expect to receive a lot of hovers for different places in that file. On the desktop when we receive an edit, we just throw away all of our caches https://sourcegraph.com/github.com/sourcegraph/go-langserver@ee6a3562/-/blob/langserver/handler.go#L353

With that in mind the major improvements for this are:
* Granular caching and cache invalidation. Do not throw away everything on edits.
* Cache intermediate work. Look into how we can use what go stores in `$GOPATH/pkg`.
* Possibly throw away the coarse pkg caching, and instead do the minimal amount of work for a hover, rather than doing the work for a whole pkg. This will technically make sourcegraph "slower", but I believe it will still be fast enough for users (instead of a few milliseconds, it will be a few 100 milliseconds).
* Do not use in-memory caches, instead rely on a filesystem (or make it pluggable).

Implementing this sort of stuff almost certainly will require forking and heavily modifying `golang.org/x/tools/go/loader`. I would target the j2d use case since that is the most heavily used. We can then get hover to use the j2d result to generate hover contents. At some point we had a fork of godef. Making godef use a virtual filesystem may be the easiest win. As for references it can be left alone since it works quite differently. Symbols just relies on the AST, so does not suffer the same performance problems.

Note: I have not done any proper profiling of the desktop use case, this is what my educated guess says the problem is. We may be surprised to find that its something silly like `FindPackage` being slow. If so, we could look into open sourcing the optimised `FindPackage` we use on sourcegraph.com.

**Comment by slimsag** [comment]
(Hi all! I'm taking over here where @keegancsmith had to leave off)

I've merged https://github.com/sourcegraph/go-langserver/pull/192 just now which will significantly help lower the memory consumption of the language server, by somewhere between 20-40% from my testing.

I suspect my above PR entirely fixes this issue, but I also have another change brewing that will lower the memory consumption even further, and make most hover/definition requests faster. I'll keep everyone here posted, I hope to have that change out by the end of this week or early next week (obviously this is just an estimate / timeline could change).

Feel free to give the language server another shot and provide feedback! You can update it by running `go get -u github.com/sourcegraph/go-langserver` and enabling the VS Code option once again.

**Comment by OneOfOne** [comment]
I'll test and get back to you, woo

**Comment by OneOfOne** [comment]
It seems stable around 4.3G of RAM (16.9G virtual), been running for 30 mins or so.

**Comment by OneOfOne** [comment]
It randomly happened again, it spiked to 25G and I had to ssh in from my phone to kill it. :( @slimsag

**Comment by slimsag** [comment]
Thanks for giving it a shot and providing feedback, @OneOfOne, we appreciate it! Any details you can provide would be helpful:

1. How long was it running before it happened? A few hours, multiple days, etc.
2. Did you notice if the spike occurred very rapidly or slowly over a long period of time?
3. If possible, which codebase does it occur on? Or if it is private, is there an open source project that you would guess to be about the same size?

As mentioned in my prior message, I am working on some further improvements that will make a larger impact here. I'll update here as soon as that lands and can be tested :)

**Comment by OneOfOne** [comment]
Glad to help, when it works it's really responsive and great.

1. Less than 10 minutes, which is odd because I worked on the same project for several hours yesterday and it didn't cause any issues.
2. It was very rapid, I was trying to get to the source of a function when it happened.
3. Sadly it's a private project, that includes 10-15 sub projects, has unsafe in it, the code base is about ~5mb of pure code.

**Comment by henvic** [comment]
A simular issue is still happening on my Mac (tried updating after these changes) on a small sized project (< 30k LOC). But here is more high CPU usage than memory.

After the update I could use the langserver for more time, but after a while it still happens.

**Comment by cossay** [comment]
I'm also having the same problem here on Ubuntu. High memory and CPU usage:
![screenshot from 2017-06-02 17-50-53](https://cloud.githubusercontent.com/assets/2929250/26738541/0b144494-47be-11e7-9fb1-f6218af31339.png)
![screenshot from 2017-06-02 17-52-13](https://cloud.githubusercontent.com/assets/2929250/26738556/1672f6e6-47be-11e7-87fa-b92c9c6b3b9e.png)

**Comment by slimsag** [comment]
Hi @cossay 

Your screenshot shows processes like `godef` running, but as far as I know that should not be running if you have set `"go.useLanguageServer": true,` in your VS Code settings. Can you confirm you're testing with that option turned on?

**Comment by cossay** [comment]
@slimsag, you are right. I have "go.useLanguageServer" set to "false" in my VS Code settings. I'm going to set it as you suggest and see what happens.

**Comment by slimsag** [comment]
It'd also be very helpful if anyone observing high memory or CPU usage could upload an SVG profile, see https://github.com/sourcegraph/go-langserver#profiling

**Comment by henvic** [comment]
@slimsag,

https://cl.ly/3l2y3g0a0Q0a
https://cl.ly/3D0Q390N2t2N

**Comment by henvic** [comment]
Now with almost 100% usage after ~45min:
https://cl.ly/3Y2z3x1z2a0h
https://cl.ly/201G40441R1u

**Comment by slimsag** [comment]
@henvic @OneOfOne (/cc @ramya-rao-a) I've made another significant improvement to the memory consumption issue. I believe this should solve the issue you've all observed with the memory usage growing constantly. Now, it should sit at some constant number based on how large your project is (e.g. it should idle at a few hundred MB instead of growing into multiple GB).

The CPU usage issue / slowness around responding to go to definition and hover requests is still known, I'm working on improving this now.

Please feel free to give the language server another shot and provide feedback:
1. Update: `go get -u github.com/sourcegraph/go-langserver`
2. Re-enable the VS Code option once again.
3. If you run into any issues, please [capture a profile](https://github.com/sourcegraph/go-langserver#profiling) and respond here.

**Comment by OneOfOne** [comment]
I'll give it a try and report back.

**Comment by OneOfOne** [comment]
Good news / bad news kinda thing, good news is it stabilized at 1.6gb (~5%) of ram, bad news is every time I type anything it spikes to using 800% CPU. @slimsag 

[pprof.go-langserver.samples.cpu.001.pb.gz](https://github.com/sourcegraph/go-langserver/files/1078328/pprof.go-langserver.samples.cpu.001.pb.gz)
[cpu.svg.gz](https://github.com/sourcegraph/go-langserver/files/1078376/cpu.svg.gz)

**Comment by slimsag** [comment]
Thanks for the feedback, @OneOfOne !

That is actually just good news. I am already aware of the CPU usage issue and am working on it now :) will update soon.

**Comment by slimsag** [comment]
(Copying the message I just posted in https://github.com/Microsoft/vscode-go/issues/853#issuecomment-310792440 since these were basically the same issue.)

Hi everyone ( /cc @ramya-rao-a ), this issue should be fixed now. 😃 Very significant improvements have been made to both goto definition and hover requests, and they now use a tiny fraction of the memory / CPU usage seen before.

To update, please run:

```
go get -u github.com/sourcegraph/go-langserver
```

And once again set `go.useLanguageServer` to `true` inside your settings to enable the language server.

Please feel free to provide any feedback here, and if all looks good we can close this issue out!

**Comment by fsouza** [comment]
@slimsag thanks for working on that!

I know that Go 1.9 is still in beta, but could you add support for it? Installing go-langserver on go 1.9beta1 fails:

```
% gimme 1.9beta1
% go get -u github.com/sourcegraph/go-langserver
# github.com/sourcegraph/go-langserver/langserver/internal/godef/go/parser
src/github.com/sourcegraph/go-langserver/langserver/internal/godef/go/parser/parser.go:2036:36: cannot use nil as type token.Pos in field value
src/github.com/sourcegraph/go-langserver/langserver/internal/godef/go/parser/parser.go:2036:42: cannot use p.lineComment (type *ast.CommentGroup) as type ast.Expr in field value:
	*ast.CommentGroup does not implement ast.Expr (missing ast.exprNode method)
src/github.com/sourcegraph/go-langserver/langserver/internal/godef/go/parser/parser.go:2036:42: too few values in struct initializer
```

**Comment by slimsag** [comment]
Hi @fsouza, thanks for the feedback. I just fixed that in https://github.com/sourcegraph/go-langserver/commit/66ff4f536c4fe4a5c525472820a563939a0ab00f so try giving `go get -u github.com/sourcegraph/go-langserver` another shot now.

**Comment by fsouza** [comment]
@slimsag awesome, it works! Thank you very much! :D

**Comment by ramya-rao-a** [comment]
@OneOfOne @henvic @vanackere @cossay Can you update the language server `go get -u github.com/sourcegraph/go-langserver`, turn the setting back on and let us know how it works for you?

**Comment by OneOfOne** [comment]
3 hours so far and it's steady at 1.5g virt mem and 0.7-1.5% cpu!

**Comment by OneOfOne** [comment]
Sadly after a while the cpu usage spikes to 40% and stays there even when idle.

**Comment by slimsag** [comment]
Hi @OneOfOne -- please do the following to give me some more insight:

1. Capture a CPU profile: `go tool pprof -svg $GOPATH/bin/go-langserver http://localhost:6060/debug/pprof/profile > cpu.svg` and upload the cpu.svg file (or email to stephen@sourcegraph.com)
2. While it is using 40% CPU, kill the process via terminal with `kill -SIGQUIT $(pgrep go-langserver)`, then inside of VS Code `Output` tab (press ctrl+backtick then click 'Output' tab, [like e.g. this](https://user-images.githubusercontent.com/3173176/27556936-8e9ce076-5a6c-11e7-955b-f106a9f46368.png)) please copy entire log output here (or email).
3. Did you activate 'Find All References', or just doing hovers and Goto / Peek definition?

**Comment by OneOfOne** [comment]
Ok I can't reproduce it anymore, so it seems to be working good, if I run into it I'll update this issue.

Thank you for your hard work @slimsag.

**Comment by slimsag** [comment]
Alright :) Please don't hesitate to open a new issue if you run into it again (even if you have little or no info)!

**Comment by slimsag** [comment]
Closing since this issue is stale / no further complaints about performance have been made. If anyone runs into issues, please do not hesitate to create a new issue! :)
